### ISSUE TITLE
So, I think I need a better way to extend states ..... (?)

### ISSUE THREAD

**Issue by zerthimon** [saltstack/salt#43725]
This may look like a weird request, but I'm having a hard time figuring out how to extend a docker_container.running state.
I've got a EC2 machine with one IP address facing the internet, which is used by nginx running in a docker container:
```
nginx-container:
  docker_container.running:
    - name: nginx
    - image: nginx
    - ports:
      - 80
    - port_bindings:
      - 80:80
    - restart_policy: always
    - binds:
      - /opt/docker/services/nginx/nginx.conf:/etc/nginx/nginx.conf:ro
      - /opt/docker/services/nginx/default.conf:/etc/nginx/conf.d/default.conf:ro
    - volumes_from:
      - main_web_site_php
    - links:
      - main_web_site_php:main_web_site_php
    - watch:
      - file: main_nginx_config
      - file: default_nginx_host
      - docker: main_web_site_container
```
This state manages an nginx container which proxies to a main_web_site_php (php-fpm) container, serves static files and uses configs from volumes it exports.

Now I would like to add more php web sites (just like the main one) and I need some way to reuse the nginx docker state. I want to write states for every additional sites, they should **include** the main nginx state and **extend** it.

In each website's sls I'd like to do this:
```
include:
  - main_nginx

extend:
  nginx-container:
    docker_container:
       - volumes_from:
         - another_web_site_php
       - links:
         - another_web_site_php:another_web_site_php
       - watch:
         - docker: another_web_site_container
```
I doubt that this will work, as I read in the docs that extend doesn't add configuration, but replaces it.
Thanks for your time reading this and I'd appreciate any help on the matter.

**Comment by yagnik** [comment]
@zerthimon you are correct only watch and require are merged, everything else is overwritten. 
There are two ways we solved this problem in the past:
- Use jinja import or include to re use the module. In this case that would cause multiple docker containers to spawn which is likely not what you want
- Use pillars in nginx template and then set those pillar on the machine.

Hope that helps :)

**Comment by The-Loeki** [comment]
Second the jinja macro / import; that's our goto as well.

Another idea would be to manage a `systemd` unit with an accumulator instead https://docs.saltstack.com/en/latest/ref/states/all/salt.states.file.html#salt.states.file.accumulated

**Comment by Ch3LL** [comment]
ping @zerthimon do their suggestions help resolve the issue?

**Comment by yagnik** [comment]
TIL about https://docs.saltstack.com/en/latest/ref/states/all/salt.states.file.html#salt.states.file.accumulated that looks like a nice way to handle this !

**Comment by zerthimon** [comment]
@yagnik @The-Loeki Thanks for your replies guys! Are you suggesting that I should macro the creation of proxy nginx containers ? In that case I'd end up with multiple nginx containers trying to listen on the same IP and port.
Regarding the systemd unit: well, I see how it can be done, but I wouldn't not really need the docker_container.* module at all. Also, I'm running Trusty which doesn't have systemd yet.

@Ch3LL Not really. I think that salt is lacking flexibility in extending states. The accumulator was mentioned earlier, it's a great instrument for templates. It would be awesome if I could use accumulator in states.

**Comment by zerthimon** [comment]
Accumulator for states... hmm.. wouldn't that be a nice feature request ?

**Comment by disaster123** [comment]
What about a special syntax in general? something like:
extend:
   nginx-container+:
    docker_container+:
       - volumes_from+:
         - another_web_site_php

+ at the end of the var means merge? instead if replace?

**Comment by The-Loeki** [comment]
@zerthimon You can do exactly the same w/Upstart, probably even easier than in SystemD. 

And while I agree there's probably some use case for extending in-place, yours ain't it; it's easily solvable through salt's other many mechanisms. Also, I wouldn't hold my breath for it, as implementing it would be extremely hairy what with all the dependency/include/extend munging already going on across formulae.

The way I read your use case:
* Create pillar stuff:

```yaml
vhosting:
  host_dir: /srv/www
  apps: /srv/php-fpm

site1:
  fqdn: site1.example.com
```  

basically you'd go like:

```jinja
# Drive it through the Pillar data; this is usually the best way to go IMHO
my-awesome-ngnix:
   docker.container:
    volumes:
{%- for site, data in pillar.sites.items() %}
      - {{ pillar.vhost.host_dir|path_join(site) }}
      - {{ pillar.vhost.app_dir|path_join(site) }}
{%- endfor %}

# You could however of course also just 'walk' the directory
{%-  for dir in salt.file.find(pillar.vhost.host_dir, type=d) %}
      - {{ dir }}
{%- endfor %}
```

**Comment by zerthimon** [comment]
Used pillar-base approach as @The-Loeki suggested.
Looks kinda dirty workaround (and necessary only because of the lack of flexibility of such a basic function as "extending" an existing state), but it works, so whatever.

As my problem is now solved, the issue may be closed.

Thanks everyone for your help!
